### PR TITLE
Implement RouteConfig interface introduced in Flutter 3.3

### PR DESCRIPTION
**Pull request by johnpryan**
This makes `GoRouter` implement the [RouterConfig](https://master-api.flutter.dev/flutter/widgets/RouterConfig-class.html) type added in Flutter 3.3.

### CODE REVIEW

File: packages/go_router/lib/src/router.dart [hunk 1]
```diff
@@ -114,6 +108,9 @@ class GoRouter extends ChangeNotifier with NavigatorObserver {
   late final GoRouterDelegate _routerDelegate;
   late final GoRouteInformationProvider _routeInformationProvider;
 
+  @override
+  final BackButtonDispatcher? backButtonDispatcher;
```
Comment (johnpryan) [inline packages/go_router/lib/src/router.dart:112]: Not sure if this is correct - should GoRouter be responsible for creating this object?

Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:112]: yes it needs to create a RootBackButtonDispatcher, otherwise android backbutton won't be wired up

Comment (johnpryan) [reply packages/go_router/lib/src/router.dart]: thanks, fixed.

**Review by loic-sharma — APPROVED** [review]
Very nice! 🚀

**Comment by chunhtai** [comment]
Can you rebase off the go_router_v5 to fix the ci?

File: packages/go_router/test/inherited_test.dart [hunk 2]
```diff
@@ -135,4 +135,8 @@ class MockGoRouter extends GoRouter {
       Object? extra}) {
     latestPushedName = name;
   }
+
+  @override
+  // TODO: implement backButtonDispatcher
```
Comment (chunhtai) [inline packages/go_router/test/inherited_test.dart]: This is a test, maybe remove the todo comment?

File: packages/go_router/lib/src/router.dart [hunk 3]
```diff
@@ -54,6 +57,7 @@ class GoRouter extends ChangeNotifier with NavigatorObserver {
     List<NavigatorObserver>? observers,
     bool debugLogDiagnostics = false,
     String? restorationScopeId,
+    BackButtonDispatcher? backButtonDispatcher,
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:62]: Is there a use case for this? I don't think we need to expose this api.

File: packages/go_router/lib/src/router.dart [hunk 4]
```diff
@@ -95,6 +99,10 @@ class GoRouter extends ChangeNotifier with NavigatorObserver {
         child: nav,
       ),
     );
+
+    this.backButtonDispatcher =
```
Comment (chunhtai) [inline packages/go_router/lib/src/router.dart:103]: This can be in initializer

**Comment by johnpryan** [comment]
Ready for another review

File: packages/go_router/pubspec.yaml [hunk 5]
```diff
@@ -7,7 +7,7 @@ issue_tracker: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+
 
 environment:
   sdk: ">=2.17.0 <3.0.0"
-  flutter: ">=3.1.0-0.0.pre.2106"
+  flutter: ">=3.3.0-0.5.pre"
```
Comment (chunhtai) [inline packages/go_router/pubspec.yaml:10]: I am not sure if the ci support anything above 3.1

**Review by chunhtai — COMMENTED** [review]
There are also examples in others/ that can make use of this API

File: packages/go_router/test/inherited_test.dart (hunk 2 above)
Comment (johnpryan) [reply packages/go_router/test/inherited_test.dart]: removed

File: packages/go_router/lib/src/router.dart (hunk 3 above)
Comment (johnpryan) [reply packages/go_router/lib/src/router.dart:62]: No I'm happy with defaulting to the RootBackButtonDispatcher.

**Comment by johnpryan** [comment]
Ready for another look

**Review by chunhtai — APPROVED** [review]
LGTM once the comments below are addressed.

There are some code sample in readme too, can you also update those? Also It will be good to have a simple test in go_router_test to make sure it can be used this way.

**Comment by johnpryan** [comment]
@chunhtai I went ahead and upgraded the tests to use routerConfig ([fc69855](https://github.com/flutter/packages/pull/2524/commits/fc69855a4dc3e45f52524b05c681cc09d59c9b2e))

**Pull request merged**